hype-submitter/src/main/java/com/spotify/hype/runner/KubernetesDockerRunner.java

```diff
@@ -256,38 +257,56 @@ private Pod getBasePod(RunEnvironment env, String image) {
     }
   }
 
-  private VolumeMountInfo volumeMountInfo(PersistentVolumeClaim claim, VolumeMount volumeMount) {
-    final String claimName = claim.getMetadata().getName();
-
-    final Volume volume = new VolumeBuilder()
-        .withName(claimName)
-        .withNewPersistentVolumeClaim(claimName, volumeMount.readOnly())
-        .build();
+  private VolumeMountInfo volumeMountInfo(Volume volume, VolumeMount volumeMount) {
 
     final io.fabric8.kubernetes.api.model.VolumeMount mount = new VolumeMountBuilder()
-        .withName(claimName)
+        .withName(volume.getName())
         .withMountPath(volumeMount.mountPath())
         .withReadOnly(volumeMount.readOnly())
         .build();
 
     final String ro = volumeMount.readOnly() ? "readOnly" : "readWrite";
-    LOG.info("Mounting {} {} at {}", claimName, ro, volumeMount.mountPath());
+    LOG.info("Mounting {} {} at {}", volume.getName(), ro, volumeMount.mountPath());
 
     return new VolumeMountInfoBuilder()
-        .persistentVolumeClaim(claim)
         .volume(volume)
         .volumeMount(mount)
         .build();
   }
 
-  private List<VolumeMountInfo> volumeMountInfos(List<VolumeMount> volumeMounts) {
-    final Map<VolumeRequest, PersistentVolumeClaim> claims = volumeMounts.stream()
-        .map(VolumeMount::volumeRequest)
-        .distinct()
-        .collect(toMap(identity(), volumeRepository::getClaim));
+  private Volume buildVolume(VolumeMount volumeMount) {
+    final com.spotify.hype.model.Volume volume = volumeMount.volume();
+
+    if (volume instanceof VolumeRequest) {
+      final VolumeRequest volumeRequest = (VolumeRequest) volume;
+      final PersistentVolumeClaim claim = volumeRepository.getClaim(volumeRequest);
+      final String claimName = claim.getMetadata().getName();
+      return new VolumeBuilder()
+          .withName(claimName)
+          .withNewPersistentVolumeClaim(claimName, volumeMount.readOnly())
+          .build();
+    } else if (volume instanceof PersistentDisk) {
```

Comment:
I feel like we can do better than having this if statement, which is pretty restrictive for users who want to add arbitrary volumes which are supported by kubernetes.

One idea is having a `KubernetesVolume` interface which has a method to get the kubernetes volume object. That way this code lives with the volume code instead of in the kubernetes runner and any user could implement whatever volumes to attach they want. Another benefit is that you could use that to specify which runner volumes support. For example this volume is not supported in the LocalRunner and there should be some exception thrown preventing people from trying to use it.

Another concern is that how do you allow a volume to be used for multiple runners, you would need a combining interface or something which would get hairy if we add ever more runners.